### PR TITLE
Skip accessing the Github API if cache is newer than x days, fix #386

### LINKED ISSUE #386: Failed to get helm version from github api, response code was 403
**Is this a request for help?**: yes

---

**Is this a BUG REPORT or FEATURE REQUEST?** (choose one): FEATURE REQUEST

**What happened:**
Recently, we have been frequently encountering the error:

"Failed to get helm version from GitHub API, response code was 403"

This error occurs when the plugin tries to fetch the latest version of Helm, but GitHub has a rate limit of 60 requests per hour for the endpoint "https://api.github.com/repos/helm/helm/releases/latest".

Caching for the call is implemented as follows: cache.map(ReleaseCache::getEtag).ifPresent(etag -> connection.setRequestProperty("If-None-Match", etag));

The eTag is retrieved from the cache, and each time a request is made to GitHub to check if the resources have changed.

**What you expected to happen**:
However, the suggestion is to improve this by checking whether the cache is valid. If it is valid, the cache should be returned without making the API call. The validity of the cache could be configurable.

**How to reproduce it** (as minimally and precisely as possible):
Refresh the address https://api.github.com/repos/helm/helm/releases/latest from the browser approximately 60 times until the rate limit is reached in the response. Then, build the project with the plugin, and you will encounter the aforementioned error.

### PR DESCRIPTION
**Pull request by sschnabe**

### CODE REVIEW

**Review by nbykowski — CHANGES_REQUESTED** [review]
Unfortunately, the tests need to be fixed because the project doesn't build

File: src/main/java/io/kokuwa/maven/helm/github/Github.java [hunk 1]
```diff
@@ -2,9 +2,12 @@
 
 import java.io.IOException;
 import java.net.HttpURLConnection;
+import java.net.ResponseCache;
```
[the remark this answers is not in the document]
Comment (sschnabe) [reply src/main/java/io/kokuwa/maven/helm/github/Github.java]: Thx, i've fixed this. I've run mvn verify on main branch to check :'(

File: src/main/java/io/kokuwa/maven/helm/github/Github.java [hunk 2]
```diff
@@ -92,7 +100,8 @@ private Optional<ReleaseCache> readCache() {
 
 		try {
 			ReleaseCache cache = mapper.readValue(file.toFile(), ReleaseCache.class);
-			log.debug("Github cache found at " + file);
+			cache.setTimestamp(Files.getLastModifiedTime(file).toInstant());
```
Comment (slawekjaranowski) [inline src/main/java/io/kokuwa/maven/helm/github/Github.java:103]: I would like to set this date as time of creation a cache during cache create.

Somebody can store cache file in shared resources, which can be copied between nodes, so file modification time can change.

Comment (sschnabe) [reply src/main/java/io/kokuwa/maven/helm/github/Github.java:103]: Good point. I will include the timestamp in json file.

Comment (sschnabe) [reply src/main/java/io/kokuwa/maven/helm/github/Github.java]: Rafactored, can you check again?

**Review by slawekjaranowski — APPROVED** [review]
looks ok

File: src/main/java/io/kokuwa/maven/helm/AbstractHelmMojo.java [hunk 3]
```diff
@@ -133,6 +133,14 @@ public abstract class AbstractHelmMojo extends AbstractChartDirectoryMojo {
 	@Parameter(property = "helm.githubUserAgent", defaultValue = "kokuwaio/helm-maven-plugin")
 	private String githubUserAgent;
 
+	/*
+	 * Skip accessing the Github API for helm version if cache is newer than given days.
+	 *
+	 * @since 6.16.1
```
Comment (slawekjaranowski) [inline src/main/java/io/kokuwa/maven/helm/AbstractHelmMojo.java:139]: please to consider a `6.17.0` 😄

**Review by fabian-schlegel — APPROVED** [review]

File: src/main/java/io/kokuwa/maven/helm/AbstractHelmMojo.java (hunk 3 above)
Comment (sschnabe) [reply src/main/java/io/kokuwa/maven/helm/AbstractHelmMojo.java]: Done.

**Pull request merged**